gadgets/safebooru-waifu/src/safebooru.rs

```diff
@@ -0,0 +1,25 @@
+use std::io::Error;
+
+use egui_extras::RetainedImage;
+use serde::{Deserialize, Serialize};
+
+const QUERY_URL: &str = "https://safebooru.donmai.us/posts/random.json?tags=1girl";
+
+#[derive(Serialize, Deserialize)]
+struct SafebooruResponse {
+    pub file_url: String,
+}
+
+pub async fn query_random_image() -> anyhow::Result<RetainedImage> {
+    let response = reqwest::get(QUERY_URL).await?.text().await?;
+    let response: SafebooruResponse = serde_json::from_str(&response)?;
+    let image_response = reqwest::get(response.file_url).await?;
+
+    let image = RetainedImage::from_image_bytes(
+        image_response.url().to_string(),
+        &image_response.bytes().await?,
+    )
+    .map_err(|e| Error::new(std::io::ErrorKind::Other, e))?;
```

Comment:
Why's this needed?

Reply:
`from_image_bytes()` returns a `Result<RetainedImage, String>`, and `String` does not implement `StdError` sadly

Reply:
That's an unfortunate API, but I see now. I'm not happy about us using io::Error for this though, what about something like:

```rs
// RetainedImage::from_image_bytes returns a Result<RetainedImage, String>, so we make our own error for anyhow interop
struct RetainedImageError(String);

impl Display for RetainedImageError { ... }
impl Error for RetainedImageError {}
```